corehq/util/couch_helpers.py

```diff
@@ -169,6 +171,153 @@ def paginate_view(db, view_name, chunk_size,
             view_kwargs['skip'] = 1
 
 
+class ResumableDocsByTypeIterator(object):
+    """Perform one-time resumable iteration over documents by type
+
+    Iteration can be efficiently stopped and resumed. The iteration may
+    omit documents that are added after the iteration begins or resumes
+    and may include deleted documents.
+
+    :param db: Couchdb database.
+    :param doc_types: A list of doc type names to iterate on.
+    :param iteration_key: A unique key identifying the iteration. This
+    key will be used in combination with `doc_types` to maintain state
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param chunk_size: Number of documents to yield before updating the
+    iteration checkpoint. In the worst case about this many documents
+    that were previously yielded may be yielded again if the iteration
+    is stopped and later resumed.
+    """
+
+    def __init__(self, db, doc_types, iteration_key, chunk_size=100):
+        if isinstance(doc_types, str):
+            raise TypeError("expected list of strings, got %r" % (doc_types,))
+        self.db = db
+        self.original_doc_types = doc_types = sorted(doc_types)
+        self.iteration_key = iteration_key
+        self.chunk_size = chunk_size
+        iteration_name = "{}/{}".format(iteration_key, " ".join(doc_types))
+        self.iteration_id = hashlib.sha1(iteration_name).hexdigest()
+        try:
+            self.state = db.get(self.iteration_id)
+        except ResourceNotFound:
+            # new iteration
+            self.state = {
+                "_id": self.iteration_id,
+                "doc_type": "ResumableDocsByTypeIteratorState",
+                "retry": {},
+
+                # for humans
+                "name": iteration_name,
+                "timestamp": datetime.datetime.utcnow().isoformat()
+            }
```

Comment:
I wasn't thinking about reducing code. Using JsonObject would make it clear
what the schema for this dict is and make working with it easier.

On 27 May 2016 at 14:29, Daniel Miller <notifications@github.com> wrote:

> In corehq/util/couch_helpers.py
> <https://github.com/dimagi/commcare-hq/pull/11792#discussion_r64897225>:
>
> > +        self.chunk_size = chunk_size
> > +        iteration_name = "{}/{}".format(iteration_key, " ".join(doc_types))
> > +        self.iteration_id = hashlib.sha1(iteration_name).hexdigest()
> > +        try:
> > +            self.state = db.get(self.iteration_id)
> > +        except ResourceNotFound:
> > +            # new iteration
> > +            self.state = {
> > +                "_id": self.iteration_id,
> > +                "doc_type": "ResumableDocsByTypeIteratorState",
> > +                "retry": {},
> > +
> > +                # for humans
> > +                "name": iteration_name,
> > +                "timestamp": datetime.datetime.utcnow().isoformat()
> > +            }
>
> I had considered it, but I decided to do this to minimize dependencies. I
> don't think using JsonObject would reduce the amount of code much if at
> all.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/11792/files/c110713a9e9c98d7ac6793275af103f2941f1987#r64897225>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAPPBq-DBLHubl3b5FCco195JHHtoEmSks5qFuOTgaJpZM4In7ch>
> .
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa